File: src/main/java/me/miki/shindo/ShindoAPI.java
package me.miki.shindo;

import lombok.Getter;
import me.miki.shindo.api.roles.RoleManager;
import me.miki.shindo.api.websocket.AccountType;
import me.miki.shindo.api.websocket.ShindoWebsocket;
import me.miki.shindo.api.websocket.WsIdentity;
import me.miki.shindo.api.websocket.presence.PresenceTracker;
import me.miki.shindo.gui.mainmenu.GuiShindoMainMenu;
import me.miki.shindo.gui.modmenu.GuiModMenu;
import me.miki.shindo.logger.ShindoLogger;
import me.miki.shindo.management.file.FileManager;
import net.minecraft.client.Minecraft;

import java.io.File;
import java.lang.reflect.Method;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.UUID;
import java.util.function.Supplier;

public class ShindoAPI {

    private final RoleManager roleManager = new RoleManager();
    private final PresenceTracker presence = new PresenceTracker();

    @Getter
    private final File firstLoginFile;

    @Getter
    private long launchTime;

    @Getter
    private GuiModMenu modMenu;

    @Getter
    private GuiShindoMainMenu mainMenu;

    @Getter
    private ShindoWebsocket ws;


    public ShindoAPI() {

        FileManager fileManager = Shindo.getInstance().getFileManager();
        firstLoginFile = new File(fileManager.getCacheDir(), "first.tmp");
    }

    public void init() {
        launchTime = System.currentTimeMillis();
        modMenu = new GuiModMenu();
        mainMenu = new GuiShindoMainMenu();
    }

    public void start() {
        // Suppliers sempre leem o estado ATUAL (session/account), então não precisamos recriar o bootstrap a cada troca
        Supplier<String> uuidSup = () -> {
            try {
                UUID id = Minecraft.getMinecraft().getSession().getProfile().getId();
                return id != null ? id.toString() : "";
            } catch (Exception e) {
                return "";
            }
        };

        Supplier<String> nameSup = () -> {
            try {
                return Minecraft.getMinecraft().getSession().getUsername();
            } catch (Exception e) {
                return "";
            }
        };

        Supplier<String> typeSup = () -> {
            try {
                Minecraft mc = Minecraft.getMinecraft();
                if (mc == null || mc.getSession() == null) {
                    return "LOCAL";
                }
                
                UUID uuid = mc.getSession().getProfile().getId();
                
                // Sem UUID válido = LOCAL
                if (uuid == null) {
                    return "LOCAL";
                }
                
                // Tenta detectar Microsoft account via reflection (compatível com diferentes versões)
                try {
                    Object session = mc.getSession();
                    Method getSessionType = session.getClass().getMethod("getSessionType");
                    String sessionType = (String) getSessionType.invoke(session);
                    if (sessionType != null && (sessionType.equals("msa") || sessionType.contains("microsoft"))) {
                        return "MICROSOFT";
                    }
                } catch (Exception e) {
                    ShindoLogger.error("An error occurred while trying to get the session type", e);
                }
                
                // Verifica se tem token (Microsoft accounts geralmente têm token)
                try {
                    Object session = mc.getSession();
                    Method getToken = session.getClass().getMethod("getToken");
                    String token = (String) getToken.invoke(session);
                    if (token != null && !token.isEmpty()) {
                        // Se tem token e UUID válido, provavelmente é Microsoft
                        return "MICROSOFT";
                    }
                } catch (Exception e) {
                    ShindoLogger.error("An error occurred while trying to get the token", e);
                }
                
                // Se tem UUID válido mas não é Microsoft = OFFLINE
                return "OFFLINE";
            } catch (Exception e) {
                return "LOCAL";
            }
        };

        ws = new ShindoWebsocket(URI.create("wss://ws.shindoclient.com/websocket"), true, presence);
        ws.setRoleManager(roleManager);
        ws.setProvider(() -> {
            String rawUuid = safeTrim(uuidSup.get());
            String rawName = safeTrim(nameSup.get());
            if (rawName.isEmpty()) {
                rawName = "Player";
            }

            UUID parsed = safeUUID(rawUuid);
            String effectiveUuid = rawUuid;
            AccountType accountType = AccountType.from(typeSup.get());

            // Se não temos UUID válido (ex.: ambiente dev sem launcher),
            // geramos um UUID offline estável baseado no nome e marcamos como LOCAL.
            if (parsed == null) {
                effectiveUuid = generateOfflineUuid(rawName);
                accountType = AccountType.LOCAL;
            }

            return new WsIdentity(
                    effectiveUuid,
                    rawName,
                    null,
                    accountType
            );
        });

        ws.connect();
    }

    public void stop() {
        if (ws != null) {
            ws.disconnect();
            ws = null;
        }
    }

    public void createFirstLoginFile() {
        Shindo.getInstance().getFileManager().createFile(firstLoginFile);
    }

    public boolean isFirstLogin() {
        return !firstLoginFile.exists();
    }

    // ===== Helpers de identidade =====

    private static String safeTrim(String value) {
        return value == null ? "" : value.trim();
    }

    private static UUID safeUUID(String value) {
        try {
            return (value == null || value.isEmpty()) ? null : UUID.fromString(value);
        } catch (Exception ignored) {
            return null;
        }
    }

    private static String generateOfflineUuid(String name) {
        String baseName = (name == null || name.isEmpty()) ? "Player" : name;
        return UUID.nameUUIDFromBytes(("OfflinePlayer:" + baseName).getBytes(StandardCharsets.UTF_8)).toString();
    }
}

File: src/main/kotlin/me/miki/shindo/ShindoAPI.kt
package me.miki.shindo

import com.google.gson.JsonObject
import me.miki.shindo.api.roles.RoleManager
import me.miki.shindo.api.websocket.AccountType
import me.miki.shindo.api.websocket.ShindoWebsocket
import me.miki.shindo.api.websocket.WsIdentity
import me.miki.shindo.api.websocket.message.MessageType
import me.miki.shindo.api.websocket.presence.PresenceTracker
import me.miki.shindo.gui.GuiNavigationHub
import me.miki.shindo.gui.mainmenu.GuiShindoMainMenu
import me.miki.shindo.gui.modmenu.v2.GuiModMenu
import me.miki.shindo.management.file.FileManager
import me.miki.shindo.utils.AccountUtil
import net.minecraft.client.Minecraft
import java.io.File
import java.nio.charset.StandardCharsets
import java.util.UUID
import java.util.function.BiConsumer

@Suppress("unused", "RedundantSamConstructor")
class ShindoAPI {
    private val roleManager = RoleManager()
    val presence = PresenceTracker()

    private val firstLoginFile: File
    var launchTime: Long = 0
        private set
    lateinit var modMenu: GuiModMenu
        private set
    lateinit var mainMenu: GuiShindoMainMenu
        private set
    lateinit var navigationHub: GuiNavigationHub
        private set

    var ws: ShindoWebsocket? = null
        private set

    init {
        val fileManager: FileManager = Shindo.getInstance().getFileManager()
        firstLoginFile = File(fileManager.cacheDir, "first.tmp")
    }

    fun init() {
        launchTime = System.currentTimeMillis()
        modMenu = GuiModMenu()
        mainMenu = GuiShindoMainMenu()
        navigationHub = GuiNavigationHub()
    }

    fun start() {
        ws =
            ShindoWebsocket("wss://ws.shindoclient.com/websocket", presence).apply {
                roleManager = this@ShindoAPI.roleManager
                provider =
                    object : ShindoWebsocket.IdentityProvider {
                        override fun player(): WsIdentity = buildIdentity()
                    }

                messageHandler.addObserver(
                    BiConsumer<MessageType, JsonObject?> { type, payload ->
                        Shindo.getInstance().getProfileShareManager().handleMessage(type, payload)
                    },
                )
                messageHandler.addObserver(
                    BiConsumer<MessageType, JsonObject?> { type, payload ->
                        Shindo.getInstance().getChatManager().handleMessage(type, payload)
                    },
                )
                messageHandler.addObserver(
                    BiConsumer<MessageType, JsonObject?> { type, payload ->
                        Shindo.getInstance().getBroadcastManager().handleMessage(type, payload)
                    },
                )

                connect()
            }
    }

    fun stop() {
        ws?.disconnect()
        ws = null
    }

    fun createFirstLoginFile() {
        Shindo.getInstance().getFileManager().createFile(firstLoginFile)
    }

    fun isFirstLogin(): Boolean = !firstLoginFile.exists()

    fun getEffectiveUuid(): UUID = resolveEffectiveUuid(Minecraft.getMinecraft())

    fun getAccountType(): AccountType = resolveAccountType(Minecraft.getMinecraft())

    private fun buildIdentity(): WsIdentity {
        val mc = Minecraft.getMinecraft()
        val session = mc.session

        val name = session?.username?.trim()?.takeIf { it.isNotEmpty() } ?: "Player"

        // Profile UUID is the authoritative source; fall back to entity UUID.
        val profileUuid: UUID? = session?.profile?.id ?: mc.thePlayer?.uniqueID

        val accountType = AccountUtil.detectAccountTypeFromUuid(profileUuid)

        val uuidString =
            profileUuid?.toString()
                ?: UUID
                    .nameUUIDFromBytes(
                        ("OfflinePlayer:$name").toByteArray(StandardCharsets.UTF_8),
                    ).toString()

        return WsIdentity(uuidString, name, null, accountType)
    }

    private fun resolveEffectiveUuid(mc: Minecraft): UUID {
        mc.session
            ?.profile
            ?.id
            ?.let { return it }
        mc.thePlayer?.uniqueID?.let { return it }
        val name = mc.session?.username ?: "Player"
        return UUID.nameUUIDFromBytes(
            ("OfflinePlayer:$name").toByteArray(StandardCharsets.UTF_8),
        )
    }

    private fun resolveAccountType(mc: Minecraft): AccountType = AccountUtil.detectAccountTypeFromUuid(resolveEffectiveUuid(mc))
}
